pkg/controller/certificatesigningrequest/helpers.go

```diff
@@ -80,8 +79,8 @@ func isNodeServingCert(csr *capi.CertificateSigningRequest, x509cr *x509.Certifi
 		log.Println("Usage does not match")
 		return false
 	}
-	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
-		log.Printf("CN does not match: %s\n", x509cr.Subject.CommonName)
+	if csr.Spec.Username != x509cr.Subject.CommonName {
```

Comment:
Shouldn't we do both checks, this new `CN==requestor` plus the old which checks that the requestor is actually a node?

With this new proposed way, I (as user `jussi`) could very well craft a CSR with `CN=jussi`, since `csr.Spec.Username=jussi` and the `CN:jussi` and rubber stamp would happily stamp it.

Reply:
hmm, thinking further, should the check be something like:
```
if csr.Spec.Username != x509cr.Subject.CommonName || !strings.HasPrefix(csr.Spec.Username, "system:node:") 
```
?

Reply:
Hmm that's a good point, I was mostly concerned with the possibility that a malicious actor could pretend to be a kubelet but I'll update the PR to keep the scope tight. I do wonder if we should also validate not just that there are ip addresses and host names, but that they actually correspond to the Node's addresses. What do you think?

As for the boolean operation, I think it should be `&&` otherwise it would be effectively the same as it was before. I'll probably just update the PR to perform the checks in sequence.

Reply:
> As for the boolean operation, I think it should be && otherwise it would be effectively the same as it was before. I'll probably just update the PR to perform the checks in sequence.

Actually I think you're right with respect to `||`, sorry about that.

Reply:
>  I was mostly concerned with the possibility that a malicious actor could pretend to be a kubelet

Yes, that's a good thing always to think. But AFAIK we cannot make this anyway bullet-proof at rubber stamp (or anywhere else for that matter). I (or someone else as a cluster admin) could very well craft a CSR with something like: `kubectl --as="system:node:foobar" create -f <path to faked kubelet CSR>`. That would make the `csr.Spec.Username` to be the nodes identity in the api. But that would require wider access to the API with impersonation capability, so probably at that point kubelet certs is low on the list of concerns. :)

>  I do wonder if we should also validate not just that there are ip addresses and host names, but that they actually correspond to the Node's addresses. 

That's an excellent point, I'm not sure if that's actually possible though. In many cases there's some cloud controller that actually manages the nodes addresses, so it might be that such a controller has not yet updated the addresses properly when kubelet creates the CSR. At least it would make this prone to timing issues. Could be possible in the rubber-stamp controller to "re-queue" the CSR object so that it would be re-checked. Maybe we should put that as a separate issue and handle as separate PR if we can figure out it's actually safe in all cases.

Reply:
> I'll probably just update the PR to perform the checks in sequence.

That might be actually better if anyone needs to understand from the logs why some cert was not handled.